cmd/src/actions_exec.go

```diff
@@ -586,38 +590,116 @@ fragment repositoryFields on Repository {
 
 		// Skip repos from unsupported code hosts but don't report them explicitly.
 		if !includeUnsupported && strings.ToLower(repo.ExternalRepository.ServiceType) != "github" && strings.ToLower(repo.ExternalRepository.ServiceType) != "bitbucketserver" {
-			unsupported = append(unsupported, repo.Name)
+			unsupported[repo.Name] = struct{}{}
 			continue
 		}
 
-		if repo.DefaultBranch == nil || repo.DefaultBranch.Name == "" {
-			skipped = append(skipped, repo.Name)
-			continue
+		if _, ok := reposByID[repo.ID]; !ok {
+			reposByID[repo.ID] = &ActionRepo{
+				ID:   repo.ID,
+				Name: repo.Name,
+			}
 		}
+	}
 
-		if repo.DefaultBranch.Target.OID == "" {
-			skipped = append(skipped, repo.Name)
-			continue
+	reposQueryTemplate := `{
+  repositories(names: [%s]) {
+    nodes {
+      id
+      name
+      defaultBranch {
+        name
+        target {
+          oid
+        }
+      }
+    }
+  }
+}
+
+`
+
+	skipped := make(map[string]struct{})
+
+	// This helper function fetches a subset of repos by name and is used lower down
+	// so that we query in batches.
+	fetchRepos := func(names []string) error {
+		reposQuery := fmt.Sprintf(reposQueryTemplate, strings.Join(names, ","))
+
+		var repoResult struct {
+			Data struct {
+				Repositories struct {
+					Nodes []struct {
+						ID            string `json:"id"`
+						Name          string `json:"name"`
+						DefaultBranch *struct {
+							Name   string `json:"name"`
+							Target struct {
+								OID string `json:"oid"`
+							} `json:"target"`
+						} `json:"defaultBranch"`
+					} `json:"nodes"`
+				} `json:"repositories"`
+			} `json:"data,omitempty"`
+
+			Errors []struct {
+				Message string
+				Path    []interface{}
+			} `json:"errors,omitempty"`
+		}
+
+		err = doRequest(reposQuery, map[string]interface{}{}, &repoResult)
+		if err != nil {
+			return err
 		}
 
-		if _, ok := reposByID[repo.ID]; !ok {
-			reposByID[repo.ID] = ActionRepo{
-				ID:      repo.ID,
-				Name:    repo.Name,
-				Rev:     repo.DefaultBranch.Target.OID,
-				BaseRef: repo.DefaultBranch.Name,
+		for _, repo := range repoResult.Data.Repositories.Nodes {
+			if repo.DefaultBranch == nil || repo.DefaultBranch.Name == "" {
+				skipped[repo.Name] = struct{}{}
+				continue
+			}
+			if repo.DefaultBranch.Target.OID == "" {
+				skipped[repo.Name] = struct{}{}
+				continue
+			}
+			existing, ok := reposByID[repo.ID]
+			if !ok {
+				continue
 			}
+			existing.Rev = repo.DefaultBranch.Target.OID
+			existing.BaseRef = repo.DefaultBranch.Name
+		}
+
+		return nil
+	}
+
+	repoNames := make([]string, 0, len(reposByID))
+	for _, r := range reposByID {
+		repoNames = append(repoNames, `"`+r.Name+`"`)
```

Comment:
Nitpicky, but I personally to keep the number of quotes per line to a minimum 😄 

```suggestion
		repoNames = append(repoNames, fmt.Sprintf("%q", r.Name))
```

Reply:
We need double quotes here because it's used as a parameter for a GraphQL query

Reply:
Yeah, I know and `%q` gives you those? Or am I missing something?